clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -201,359 +224,382 @@ std::set<ConstraintVariable *>
-      /* Possible pointer arithmetic: Could be LHS or RHS */
-      case BO_Add:
-      case BO_Sub:
-        if (BO->getLHS()->getType()->isPointerType())
-          return getExprConstraintVars(BO->getLHS());
-        else if (BO->getRHS()->getType()->isPointerType())
-          return getExprConstraintVars(BO->getRHS());
-        else
-          return PVConstraintFromType(TypE);
-      /* Pointer-to-member ops unsupported */
-      case BO_PtrMemD:
-      case BO_PtrMemI:
-        assert(false && "Bogus pointer-to-member operator");
-        break;
-      /* bit-shift/arithmetic/assign/comp operators return ints; do nothing */
-      case BO_ShlAssign:
-      case BO_ShrAssign:
-      case BO_AndAssign:
-      case BO_XorAssign:
-      case BO_OrAssign:
-      case BO_MulAssign:
-      case BO_DivAssign:
-      case BO_RemAssign:
-      case BO_And:
-      case BO_Or:
-      case BO_Mul:
-      case BO_Div:
-      case BO_Rem:
-      case BO_Xor:
-      case BO_Cmp:
-      case BO_EQ:
-      case BO_NE:
-      case BO_GE:
-      case BO_GT:
-      case BO_LE:
-      case BO_LT:
-      case BO_LAnd:
-      case BO_LOr:
-      case BO_Shl:
-      case BO_Shr:
-        return PVConstraintFromType(TypE);
-      }
-
-    // x[e]
-    } else if (ArraySubscriptExpr *AE = dyn_cast<ArraySubscriptExpr>(E)) {
-      std::set<ConstraintVariable *> T = getExprConstraintVars(AE->getBase());
-      std::set<ConstraintVariable *> Tmp = handleDeref(T);
-      T.swap(Tmp);
-      return T;
-
-    // ++e, &e, *e, etc.
-    } else if (UnaryOperator *UO = dyn_cast<UnaryOperator>(E)) {
-      Expr *UOExpr = UO->getSubExpr();
-      switch (UO->getOpcode()) {
-      // &e
-      // C99 6.5.3.2: "The operand of the unary & operator shall be either a
-      // function designator, the result of a [] or unary * operator, or an
-      // lvalue that designates an object that is not a bit-field and is not
-      // declared with the register storage-class specifier."
-      case UO_AddrOf: {
-        UOExpr = UOExpr->IgnoreParenImpCasts();
-        // Taking the address of a dereference is a NoOp, so the constraint
-        // vars for the subexpression can be passed through.
-        // FIXME: We've dumped implicit casts on UOEXpr; restore?
-        if (UnaryOperator *SubUO = dyn_cast<UnaryOperator>(UOExpr)) {
-          if (SubUO->getOpcode() == UO_Deref)
-            return getExprConstraintVars(SubUO->getSubExpr());
-          // else, fall through
-        } else if (ArraySubscriptExpr *ASE = dyn_cast<ArraySubscriptExpr>(UOExpr)) {
-          return getExprConstraintVars(ASE->getBase());
+      // Checked-C temporary
+    } else if (CHKCBindTemporaryExpr *CE = dyn_cast<CHKCBindTemporaryExpr>(E)) {
+      return getExprConstraintVars(CE->getSubExpr());
+    } else {
+      // Apart from the above expressions constraints for all the other
+      // expressions can be cached.
+      auto PersistentExprHandler = [this](Expr *E) {
```

Comment:
I don't see why you need to make a handler for this since you only ever call the getPersistentConstraints etc. once. That is, just create a variable CVarSet Ret here. Then add the code to check the persistent constraints and return if found. Else run the following conditional and have it set Ret. At the very end, add Ret to the persistent constraints and return it. In essence: Just inline the call to `getOrCreatePersistentConstraints` below.